services/subnet_manager/src/main/java/com/futurewei/alcor/subnet/controller/SubnetController.java

```diff
@@ -0,0 +1,251 @@
+
+            // TODO: Create a verification framework for all resources
+            SubnetState inSubnetState = resource.getSubnet();
+            RestPreconditionsUtil.verifyResourceFound(inSubnetState.getVpcId());
+            RestPreconditionsUtil.populateResourceProjectId(inSubnetState, projectid);
+
+            this.subnetDatabaseService.addSubnet(inSubnetState);
+
+            subnetState = this.subnetDatabaseService.getBySubnetId(inSubnetState.getId());
+            if (subnetState == null) {
+                throw new ResourcePersistenceException();
+            }
+
+            // Verify VPC ID
+            VpcStateJson vpcResponse = this.subnetService.verifyVpcId(projectid, inSubnetState.getVpcId());
+            if (vpcResponse == null) {
+                throw new ResourcePersistenceException();
+            }
+
+            //Prepare Route Rule(IPv4/6) for Subnet
+            RouteWebJson routeResponse = this.subnetService.prepeareRouteRule(inSubnetState.getVpcId(), vpcResponse);
+            if (routeResponse == null) {
+                throw new ResourcePersistenceException();
+            }
+
+            //Allocate Gateway Mac
+//            MacState macState = new MacState();
+//            String portId = UUID.randomUUID().toString();
+//            macState.setProjectId(projectid);
+//            macState.setPortId(portId);
+//            macState.setVpcId(inSubnetState.getVpcId());
+//
+//            HttpEntity<MacStateJson> macRequest = new HttpEntity<>(new MacStateJson(macState));
+//            MacStateJson macResponse = restTemplate.postForObject(macUrl, macRequest, MacStateJson.class);
+//            if (macResponse == null) {
+//                throw new ResourcePersistenceException();
+//            }
+
+            // Verify/Allocate Gateway IP, subnet id, port id, subnet cidr, response:IP - unique
+//            IPState ipState = new IPState();
+//            ipState.setSubnetId(inSubnetState.getId());
+//            ipState.setPortId(portId);
+//            ipState.setSubnetCidr(inSubnetState.getCidr());
+//
+//            String ipManagerServiceUrl = ipUrl + inSubnetState.getId() + "/routes"; // for kubernetes test
+//            HttpEntity<IPStateJson> ipRequest = new HttpEntity<>(new IPStateJson(ipState));
+//            IPStateJson ipResponse = restTemplate.postForObject(ipManagerServiceUrl, ipRequest, IPStateJson.class);
+//            if (ipResponse == null) {
+//                throw new ResourcePersistenceException();
+//            }
+
+            // set up value of properties for subnetState
+            List<RouteWebObject> routes = subnetState.getRoutes();
+            if (routes == null) {
+                routes = new ArrayList<>();
+            }
+            routes.add(routeResponse.getRoute());
+            subnetState.setRoutes(routes);
+            //subnetState.setGatewayIp(ipResponse.getIpState().getIp());
+
+        } catch (ResourceNullException e) {
```

Comment:
Which operation will throw ResourceNullException?

Reply:
There also change to Exception

Reply:
Same rational as above. You want to a specific exception, with better name than "ResourceNullException", maybe "VpcNotExistException"